File: trunk/client-glib/src/api/common/ifolder.cpp
/***********************************************************************
 *  $RCSfile$
 * 
 *  Copyright (C) 2006 Novell, Inc.
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  Author(s): Boyd Timothy <btimothy@novell.com>
 *
 ***********************************************************************/

#include "../ifolder-client.h"

IFOLDER_API iFolderType
ifolder_get_type(const iFolder ifolder)
{
	return IFOLDER_TYPE_DISCONNECTED;
}

IFOLDER_API const char *ifolder_get_id(const iFolder ifolder)
{
	return NULL;
}

IFOLDER_API const char *ifolder_get_name(const iFolder ifolder)
{
	return NULL;
}

IFOLDER_API const char *ifolder_get_description(const iFolder ifolder)
{
	return NULL;
}

IFOLDER_API int ifolder_set_description(const iFolder ifolder, const char *new_description)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_owner(const iFolder ifolder, iFolderUser *owner)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_domain(const iFolder ifolder, iFolderDomain *domain)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_size(const iFolder ifolder, long *size)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_file_count(const iFolder ifolder, int *file_count)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_directory_count(const iFolder ifolder, int *directory_count)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_rights(const iFolder ifolder, iFolderMemberRights *rights)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_last_modified(const iFolder ifolder, time_t *last_modified)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API bool ifolder_is_published(const iFolder ifolder)
{
	return false;
}

IFOLDER_API bool ifolder_is_enabled(const iFolder ifolder)
{
	return false;
}

IFOLDER_API int ifolder_get_member_count(const iFolder ifolder, int *member_count)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_state(const iFolder ifolder, iFolderState *state)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_items_to_synchronize(const iFolder ifolder, int *items_to_sync)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_start_synchronization(const iFolder ifolder, bool sync_now)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_stop_synchronization(const iFolder ifolder)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_resume_synchronization(const iFolder ifolder, bool sync_now)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_members(const iFolder ifolder, const int index, const int count, iFolderEnumeration *user_enum)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_set_member_rights(const iFolder ifolder, const iFolderUser member, const iFolderMemberRights rights)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_add_member(const iFolder ifolder, const iFolderUser member, const iFolderMemberRights rights)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_remove_member(const iFolder ifolder, const iFolderUser member)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_set_owner(const iFolder ifolder, const iFolderUser member)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_get_change_entries(const iFolder ifolder, const int index, const int count, iFolderEnumeration *change_entry_enum)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_publish(const iFolder ifolder)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API int ifolder_unpublish(const iFolder ifolder)
{
	return IFOLDER_UNIMPLEMENTED;
}

IFOLDER_API void ifolder_free(iFolder ifolder)
{
}

File: src/WebService/iFolder.cs
/***********************************************************************
 *  $RCSfile$
 *
 *  Copyright (C) 2004 Novell, Inc.
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this program; if not, write to the Free
 *  Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *  Author: Calvin Gaisford <cgaisford@novell.com>
 *
 ***********************************************************************/

using System;
using Simias.Storage;
using Simias.POBox;
using Simias.Sync;
using System.Xml;
using System.Xml.Serialization;

namespace Novell.iFolder.Web
{
	/// <summary>
	/// This class exists only to represent an iFolder and should only be
	/// used in association with the iFolderWebService class.
	/// </summary>
	[Serializable]
	public class iFolderWeb
	{
		/// <summary>
		/// </summary>
		public static readonly string FilesDirName = "SimiasFiles";
		/// <summary>
		/// </summary>
		public static readonly string iFolderType = "iFolder";

		/// <summary>
		/// </summary>
		public string DomainID;
		/// <summary>
		/// </summary>
		public string ID;
		/// <summary>
		/// </summary>
		public ulong LocalIncarnation;
		/// <summary>
		/// </summary>
		public string ManagedPath;
		/// <summary>
		/// </summary>
		public string UnManagedPath;
		/// <summary>
		/// </summary>
		public ulong MasterIncarnation;
		/// <summary>
		/// </summary>
        public string Name;
		/// <summary>
		/// </summary>
		public string Owner;
		/// <summary>
		/// </summary>
		public string OwnerID;
		/// <summary>
		/// </summary>
		public int EffectiveSyncInterval;
		/// <summary>
		/// </summary>
		public int SyncInterval;
		/// <summary>
		/// </summary>
		public bool Synchronizable;
		/// <summary>
		/// </summary>
		public string Type;
		/// <summary>
		/// </summary>
		public string Description;
		/// <summary>
		/// </summary>
		public string State;
		/// <summary>
		/// </summary>
		public bool IsSubscription;
		/// <summary>
		/// </summary>
		public int EnumeratedState;
		/// <summary>
		/// </summary>
		public bool IsWorkgroup;
		/// <summary>
		/// </summary>
		public bool HasConflicts;
		/// <summary>
		/// </summary>
		public string CurrentUserID;
		/// <summary>
		/// </summary>
		public string CurrentUserRights;
		/// <summary>
		/// </summary>
		public string CollectionID;
		/// <summary>
		/// </summary>
		public string LastSyncTime;
		/// <summary>
		/// </summary>
		public string Role;

		public int encryption_status;
		public string passphrase;

		/// <summary>
		/// </summary>
		public iFolderWeb()
		{
		}

		/// <summary>
		/// </summary>
		public iFolderWeb(Collection collection)
		{
			this.DomainID = collection.Domain;
			this.ID = collection.ID;
			this.CollectionID = collection.ID;
			this.LocalIncarnation = collection.LocalIncarnation;
			DirNode dirNode = collection.GetRootDirectory();
			if(dirNode != null)
				this.UnManagedPath = dirNode.GetFullPath(collection);
			else
				this.UnManagedPath = "";
			this.ManagedPath = collection.ManagedPath;
			this.MasterIncarnation = collection.MasterIncarnation;
			this.Name = collection.Name;
			if(collection.Owner != null)
			{
				this.Owner = collection.Owner.Name;
				this.OwnerID = collection.Owner.UserID;
			}
			else
			{
				this.Owner = "Not available";
				this.OwnerID = "0";
			}

			this.SyncInterval = 
				Simias.Policy.SyncInterval.GetInterval(collection);
			this.Synchronizable = collection.Synchronizable;
			this.Type = iFolderType;
			this.Description = "";
			this.IsSubscription = false;
			this.EnumeratedState = -1;

			// There is no longer a WorkGroup domain ... for now return false. We
			// probably need to change this to return a type instead.
			this.IsWorkgroup = false;
			this.HasConflicts = collection.HasCollisions();

			Member tmpMember = collection.GetCurrentMember();
			this.CurrentUserID = tmpMember.UserID;
			this.CurrentUserRights = tmpMember.Rights.ToString();

			Simias.Policy.SyncInterval si = Simias.Policy.SyncInterval.Get(tmpMember, collection);
			this.EffectiveSyncInterval = si.Interval;
			DateTime lastSyncTime = Simias.Sync.SyncClient.GetLastSyncTime(collection.ID);
			if (collection.Role.Equals(SyncRoles.Master))
			{
				this.LastSyncTime = string.Empty;
				this.State = "Local";
			}
			else if (lastSyncTime.Equals(DateTime.MinValue))
			{
				this.LastSyncTime = string.Empty;
				this.State = "WaitSync";
			}
			else
			{
				this.LastSyncTime = lastSyncTime.ToString();
				this.State = collection.IsProxy ? "WaitSync" : "Local";
			}

			this.Role = collection.Role.ToString();
			this.encryption_status = collection.Encryption_Status;
		}


		/// <summary>
		/// </summary>
		public iFolderWeb(Subscription subscription)
		{
			this.DomainID = subscription.DomainID;
			this.Name = subscription.SubscriptionCollectionName;
			this.ID = subscription.ID;
			this.CollectionID = subscription.SubscriptionCollectionID;
			this.Description = subscription.CollectionDescription;
			this.IsSubscription = true;
			this.EnumeratedState = (int) subscription.SubscriptionState;
			this.Owner = subscription.FromName;
			Domain domain = Store.GetStore().GetDomain(subscription.DomainID);
			if(domain != null)
			{
				Simias.Storage.Member member = domain.GetMemberByID(subscription.FromIdentity);
				if(member != null)
					this.Owner = member.FN;
			}

			this.OwnerID = subscription.FromIdentity;
			this.CurrentUserRights = subscription.SubscriptionRights.ToString();

			if(	(subscription.SubscriptionState == 
								SubscriptionStates.Ready) ||
						(subscription.SubscriptionState == 
								SubscriptionStates.Received) )
			{
				this.State = "Available";
			}
			else if(subscription.SubscriptionState == 
									SubscriptionStates.Replied)
			{
				this.State = "WaitConnect";
			}
			else if(subscription.SubscriptionState == 
								SubscriptionStates.Delivered)
			{
				this.State = "WaitSync";
			}
/*			else if(	(subscription.SubscriptionState == 
								SubscriptionStates.Replied) ||
						(subscription.SubscriptionState == 
								SubscriptionStates.Delivered) ||
						(subscription.SubscriptionState == 
								SubscriptionStates.Pending) ||
						(subscription.SubscriptionState == 
								SubscriptionStates.Responded) ||
						(subscription.SubscriptionState == 
								SubscriptionStates.Acknowledged) )
*/
			else
			{
				this.State = "Unknown";
			}
		}
	}
}
